File: frontend/src/api/simulations.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api } from "./client";

export interface SimulationRunOut {
  id: string;
  name: string;
  description: string | null;
  status: "draft" | "computed";
  snapshot_date: string;
  created_by: string;
  created_at: string;
}

export interface SimulationMoveOut {
  id: string;
  simulation_run_id: string;
  employee_id: string;
  from_team_id: string;
  to_team_id: string;
  allocation_pct_change: number;
  notes: string | null;
}

export interface SimulationOutcomeOut {
  id: string;
  simulation_run_id: string;
  employee_id: string;
  metric_name: string;
  baseline_value: number | null;
  simulated_value: number;
  delta: number;
  team_scope_id: string | null;
}

export interface SimulationCreateIn {
  name: string;
  description?: string;
  snapshot_date: string;
}

export interface SimulationMoveIn {
  employee_id: string;
  from_team_id: string;
  to_team_id: string;
  allocation_pct_change?: number;
  notes?: string;
}

export function useCreateSimulation() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: SimulationCreateIn) =>
      api.post<SimulationRunOut>("/simulations/", body).then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["simulations"] });
    },
  });
}

export function useAddMove(simId: string) {
  return useMutation({
    mutationFn: (body: SimulationMoveIn) =>
      api.post<SimulationMoveOut>(`/simulations/${simId}/moves`, body).then((r) => r.data),
  });
}

export function useComputeSimulation() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (simId: string) =>
      api.post<SimulationRunOut>(`/simulations/${simId}/compute`).then((r) => r.data),
    onSuccess: (_data, simId) => {
      qc.invalidateQueries({ queryKey: ["simulation", simId, "outcomes"] });
    },
  });
}

export function useSimulationOutcomes(simId: string | null) {
  return useQuery<SimulationOutcomeOut[]>({
    queryKey: ["simulation", simId, "outcomes"],
    queryFn: () =>
      api.get(`/simulations/${simId}/outcomes`).then((r) => r.data),
    enabled: !!simId,
  });
}

File: backend/schemas/simulations.py
"""
Pydantic v2 schemas for simulation endpoints (Sprint 6).
"""

import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict


# ---------------------------------------------------------------------------
# Input schemas
# ---------------------------------------------------------------------------

class SimulationCreateIn(BaseModel):
    name: str
    description: str | None = None
    snapshot_date: date


class SimulationMoveIn(BaseModel):
    employee_id: uuid.UUID
    from_team_id: uuid.UUID
    to_team_id: uuid.UUID
    allocation_pct_change: float = 0.0
    notes: str | None = None


# ---------------------------------------------------------------------------
# Output schemas
# ---------------------------------------------------------------------------

class SimulationRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    description: str | None
    status: str
    snapshot_date: date
    created_by: str
    created_at: datetime


class SimulationMoveOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    simulation_run_id: uuid.UUID
    employee_id: uuid.UUID
    from_team_id: uuid.UUID
    to_team_id: uuid.UUID
    allocation_pct_change: Decimal
    notes: str | None


class SimulationOutcomeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    simulation_run_id: uuid.UUID
    employee_id: uuid.UUID
    metric_name: str
    baseline_value: Decimal | None
    simulated_value: Decimal
    delta: Decimal
    team_scope_id: uuid.UUID | None
